store/postgres/src/advisory_lock.rs

```diff
@@ -19,9 +19,57 @@ use diesel::{sql_query, PgConnection, RunQueryDsl};
 use graph::prelude::StoreError;
 
 use crate::command_support::catalog::Site;
+use crate::primary::DeploymentId;
 
-/// Get a lock for running migrations. Blocks until we get
-/// the lock.
+/// A locking scope for a particular deployment. We use different scopes for
+/// different purposes, and in each scope we use an advisory lock for each
+/// deployment.
+struct Scope {
+    id: i32,
+}
+
+impl Scope {
+    /// Try to lock the deployment in this scope with the given id. Return
+    /// `true` if we got the lock, and `false` if it is already locked.
+    fn try_lock(&self, conn: &PgConnection, id: DeploymentId) -> Result<bool, StoreError> {
+        #[derive(QueryableByName)]
+        struct Locked {
+            #[sql_type = "Bool"]
+            locked: bool,
+        }
+
+        sql_query(format!(
+            "select pg_try_advisory_lock({}, {id}) as locked",
+            self.id
+        ))
+        .get_result::<Locked>(conn)
+        .map(|res| res.locked)
+        .map_err(StoreError::from)
+    }
+
+    /// Lock the deployment in this scope with the given id. Blocks until we
+    /// can get the lock
+    fn lock(&self, conn: &PgConnection, id: DeploymentId) -> Result<(), StoreError> {
+        sql_query(format!("select pg_advisory_lock({}, {id})", self.id))
+            .execute(conn)
+            .map(|_| ())
+            .map_err(StoreError::from)
+    }
+
+    /// Unlock the deployment in this scope with the given id.
+    fn unlock(&self, conn: &PgConnection, id: DeploymentId) -> Result<(), StoreError> {
+        sql_query(format!("select pg_advisory_unlock({}, {id})", self.id))
+            .execute(conn)
+            .map(|_| ())
+            .map_err(StoreError::from)
+    }
+}
+
+const COPY: Scope = Scope { id: 1 };
+const WRITE: Scope = Scope { id: 2 };
+const PRUNE: Scope = Scope { id: 3 };
```

Comment:
Good question; I think it's ok if the source subgraph gets pruned simultaneously:
*  we might end up with data that's been pruned from the source, but that's harmless since the `earliest_block` of the destination will be set to that from the source _after_ data copying has finished and will make that data unreachable
* we might also try to copy data that has been pruned after we determined which data to copy, which is fine for the same reason

If the destination subgraph gets pruned while the copy is still in progress, things are ok for the raw data, but I think we might get into trouble with these steps:

* copy process starts and copies data
* pruning does its thing and sets `earliest_block`
* copying data finishes and sets metadata, including `earliest_block` from the source which might be before what was pruned

We could fix this by never allowing `earliest_block` to go back by changing `deployment::earliest_block` to
```rust
pub fn set_earliest_block(
    conn: &PgConnection,
    site: &Site,
    earliest_block: BlockNumber,
) -> Result<(), StoreError> {
    use subgraph_deployment as d;

    update(d::table.filter(d::id.eq(site.id)))
        .set(d::earliest_block_number.eq(earliest_block))
       .filter(d::earliest_block.lt(earliest_block)) // this is the change
        .execute(conn)?;
    Ok(())
}
```

Reply:
There's one more interaction between copying and pruning that I am not clear on: as things are right now, when you copy/graft a deployment that has `history_blocks` set, that setting gets reset to the default, i.e., pruning is disabled for the copy. I think copying/grafting should retain that setting.

Reply:
Nice, thanks for thinking through this, it's worth documenting some of that analysis somewhere on this file.

Reply:
>     * we might end up with data that's been pruned from the source, but that's harmless since the `earliest_block` of the destination will be set to that from the source _after_ data copying has finished and will make that data unreachable

I just looked through the code again, and that's actually not true. I filed #4496 to capture this. I'll merge this PR as-is and work on another PR to address this problem.